File: src/test/java/io/github/sparkle4j/AppcastParserTest.java
package io.github.sparkle4j;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.nio.charset.StandardCharsets;

import static org.junit.jupiter.api.Assertions.*;

class AppcastParserTest {

    private final AppcastParser parser = new AppcastParser();

    private String loadResource(String name) throws IOException {
        try (var stream = getClass().getClassLoader().getResourceAsStream(name)) {
            if (stream == null) throw new IOException("Resource not found: " + name);
            return new String(stream.readAllBytes(), StandardCharsets.UTF_8);
        }
    }

    @Test @DisplayName("parses valid appcast and returns items sorted by version descending")
    void parsesValidAppcast() throws Exception {
        var items = parser.parse(loadResource("appcast-valid.xml"));
        assertFalse(items.isEmpty());
        // Items should be sorted descending
        for (int i = 0; i < items.size() - 1; i++) {
            assertTrue(VersionComparator.isNewer(items.get(i).version(), items.get(i + 1).version())
                || items.get(i).version().equals(items.get(i + 1).version()));
        }
    }

    @Test @DisplayName("filters to current OS only")
    void filtersToCurrentOs() throws Exception {
        var items = parser.parse(loadResource("appcast-valid.xml"));
        var currentOs = AppcastParser.currentOs();
        for (var item : items) {
            assertEquals(currentOs, item.os());
        }
    }

    @Test @DisplayName("rejects non-HTTPS enclosure URLs")
    void rejectsNonHttps() throws Exception {
        var xml = """
            <?xml version="1.0" encoding="UTF-8"?>
            <rss version="2.0" xmlns:sparkle="http://www.andymatuschak.org/xml-namespaces/sparkle">
              <channel>
                <item>
                  <sparkle:version>1.0.0</sparkle:version>
                  <enclosure url="http://example.com/update.exe" sparkle:os="%s" length="100"/>
                </item>
              </channel>
            </rss>
            """.formatted(AppcastParser.currentOs());
        var items = parser.parse(xml);
        assertTrue(items.isEmpty());
    }

    @Test @DisplayName("parses critical update flag")
    void parsesCriticalUpdate() throws Exception {
        var xml = """
            <?xml version="1.0" encoding="UTF-8"?>
            <rss version="2.0" xmlns:sparkle="http://www.andymatuschak.org/xml-namespaces/sparkle">
              <channel>
                <item>
                  <sparkle:version>1.0.0</sparkle:version>
                  <sparkle:criticalUpdate/>
                  <enclosure url="https://example.com/update.exe" sparkle:os="%s" length="100"/>
                </item>
              </channel>
            </rss>
            """.formatted(AppcastParser.currentOs());
        var items = parser.parse(xml);
        assertEquals(1, items.size());
        assertTrue(items.get(0).isCriticalUpdate());
    }

    @Test @DisplayName("returns empty list for malformed XML")
    void malformedXml() {
        assertThrows(Exception.class, () -> parser.parse("not xml at all"));
    }

    @Test @DisplayName("returns empty list when no enclosure matches current OS")
    void noMatchingOs() throws Exception {
        var otherOs = AppcastParser.currentOs().equals("windows") ? "macos" : "windows";
        var xml = """
            <?xml version="1.0" encoding="UTF-8"?>
            <rss version="2.0" xmlns:sparkle="http://www.andymatuschak.org/xml-namespaces/sparkle">
              <channel>
                <item>
                  <sparkle:version>1.0.0</sparkle:version>
                  <enclosure url="https://example.com/update.exe" sparkle:os="%s" length="100"/>
                </item>
              </channel>
            </rss>
            """.formatted(otherOs);
        var items = parser.parse(xml);
        assertTrue(items.isEmpty());
    }

    @Test @DisplayName("handles missing optional fields gracefully")
    void missingOptionalFields() throws Exception {
        var xml = """
            <?xml version="1.0" encoding="UTF-8"?>
            <rss version="2.0" xmlns:sparkle="http://www.andymatuschak.org/xml-namespaces/sparkle">
              <channel>
                <item>
                  <sparkle:version>1.0.0</sparkle:version>
                  <enclosure url="https://example.com/update.exe" sparkle:os="%s" length="100"/>
                </item>
              </channel>
            </rss>
            """.formatted(AppcastParser.currentOs());
        var items = parser.parse(xml);
        assertEquals(1, items.size());
        var item = items.get(0);
        assertNull(item.edSignature());
        assertNull(item.releaseNotesUrl());
        assertNull(item.pubDate());
        assertFalse(item.isCriticalUpdate());
    }

    @Test @DisplayName("enclosure version overrides item version")
    void enclosureVersionOverridesItem() throws Exception {
        var xml = """
            <?xml version="1.0" encoding="UTF-8"?>
            <rss version="2.0" xmlns:sparkle="http://www.andymatuschak.org/xml-namespaces/sparkle">
              <channel>
                <item>
                  <sparkle:version>1.0.0</sparkle:version>
                  <enclosure url="https://example.com/update.exe" sparkle:os="%s" sparkle:version="2.0.0" length="100"/>
                </item>
              </channel>
            </rss>
            """.formatted(AppcastParser.currentOs());
        var items = parser.parse(xml);
        assertEquals("2.0.0", items.get(0).version());
    }

    @Test @DisplayName("filters by minimumJavaVersion")
    void filtersMinJava() throws Exception {
        var items = parser.parse(loadResource("appcast-min-java.xml"));
        for (var item : items) {
            if (item.minimumJavaVersion() != null) {
                assertTrue(Runtime.version().feature() >= item.minimumJavaVersion());
            }
        }
    }

    @Test @DisplayName("uses inline description when no release notes URL")
    void inlineDescription() throws Exception {
        var items = parser.parse(loadResource("appcast-inline-description.xml"));
        assertFalse(items.isEmpty());
        var item = items.get(0);
        assertNull(item.releaseNotesUrl());
        assertNotNull(item.inlineDescription());
    }

    @Test @DisplayName("shortVersionString falls back to version when absent")
    void shortVersionFallback() throws Exception {
        var xml = """
            <?xml version="1.0" encoding="UTF-8"?>
            <rss version="2.0" xmlns:sparkle="http://www.andymatuschak.org/xml-namespaces/sparkle">
              <channel>
                <item>
                  <sparkle:version>1.0.0</sparkle:version>
                  <enclosure url="https://example.com/update.exe" sparkle:os="%s" length="100"/>
                </item>
              </channel>
            </rss>
            """.formatted(AppcastParser.currentOs());
        var items = parser.parse(xml);
        assertEquals("1.0.0", items.get(0).shortVersionString());
    }

    @Test @DisplayName("default installer type is inno")
    void defaultInstallerType() throws Exception {
        var xml = """
            <?xml version="1.0" encoding="UTF-8"?>
            <rss version="2.0" xmlns:sparkle="http://www.andymatuschak.org/xml-namespaces/sparkle">
              <channel>
                <item>
                  <sparkle:version>1.0.0</sparkle:version>
                  <enclosure url="https://example.com/update.exe" sparkle:os="%s" length="100"/>
                </item>
              </channel>
            </rss>
            """.formatted(AppcastParser.currentOs());
        var items = parser.parse(xml);
        assertEquals("inno", items.get(0).installerType());
    }
}

File: src/test/kotlin/io/github/sparkle4j/AppcastParserTest.kt
package io.github.sparkle4j

import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertNull
import kotlin.test.assertTrue

class AppcastParserTest {

    private val parser = AppcastParser()

    private fun xml(resource: String): String = AppcastParserTest::class.java.classLoader
        .getResourceAsStream(resource)!!.bufferedReader().readText()

    // --- happy path ---

    @Test
    fun `parses valid feed and returns highest version first`() {
        val items = withOs("windows") { parser.parse(xml("appcast-valid.xml")) }
        assertTrue(items.isNotEmpty(), "Expected at least one item")
        // Machine-readable version (build number) used for sorting
        assertEquals("47391", items.first().version)
        // Human-readable version shown in UI
        assertEquals("0.7.0", items.first().shortVersionString)
    }

    @Test
    fun `picks correct enclosure for windows`() {
        val items = withOs("windows") { parser.parse(xml("appcast-valid.xml")) }
        val item = items.first()
        assertEquals("windows", item.os)
        assertTrue(item.url.contains("windows"), "Expected windows URL")
        assertEquals(185_000_000L, item.length)
        assertEquals("AAAA", item.edSignature)
    }

    @Test
    fun `reads criticalUpdate element`() {
        val items = withOs("windows") { parser.parse(xml("appcast-valid.xml")) }
        assertTrue(items.first().isCriticalUpdate, "0.7.0 should be marked as critical")
    }

    @Test
    fun `picks correct enclosure for linux`() {
        val items = withOs("linux") { parser.parse(xml("appcast-valid.xml")) }
        val item = items.first()
        assertEquals("linux", item.os)
        assertTrue(item.url.endsWith(".deb"))
    }

    @Test
    fun `populates optional fields`() {
        val items = withOs("windows") { parser.parse(xml("appcast-valid.xml")) }
        val item = items.first()
        assertEquals("https://example.com/changelog", item.releaseNotesUrl)
        assertNotNull(item.pubDate)
    }

    // --- filtering ---

    @Test
    fun `skips items without enclosure for current OS`() {
        // appcast-valid.xml has no macos 0_6_0 enclosure — only one macos item
        val items = withOs("macos") { parser.parse(xml("appcast-valid.xml")) }
        assertEquals(1, items.size)
        assertEquals("0.7.0", items.first().shortVersionString)
    }

    @Test
    fun `skips items whose minimum java version exceeds current JVM`() {
        val items = withOs("windows") { parser.parse(xml("appcast-min-java.xml")) }
        assertTrue(items.isEmpty(), "Item requiring Java 999 should be filtered out")
    }

    @Test
    fun `rejects non-HTTPS enclosure URL`() {
        val feedWithHttp = xml("appcast-valid.xml")
            .replace(
                "https://example.com/releases/testapp-0.7.0-windows.exe",
                "http://example.com/releases/testapp-0.7.0-windows.exe",
            )
        val items = withOs("windows") { parser.parse(feedWithHttp) }
        // 0.7.0 has no valid windows enclosure now; 0.6.0 still has one
        assertTrue(items.none { it.version == "0.7.0" }, "0.7.0 should be rejected (http URL)")
    }

    @Test
    fun `falls back to inline description when no releaseNotesLink`() {
        val items = withOs("windows") { parser.parse(xml("appcast-inline-description.xml")) }
        val item = items.first()
        assertNull(item.releaseNotesUrl)
        assertTrue(item.inlineDescription?.contains("Fixed a crash") == true)
    }

    @Test
    fun `shortVersionString falls back to version when absent`() {
        val feed = xml("appcast-valid.xml")
            .replace("<sparkle:shortVersionString>0.7.0</sparkle:shortVersionString>", "")
        val items = withOs("windows") { parser.parse(feed) }
        // shortVersionString should equal version when the element is absent
        assertEquals(items.first().version, items.first().shortVersionString)
    }

    // --- error cases ---

    @Test
    fun `throws on malformed XML`() {
        assertThrows<Exception> { parser.parse("<not valid xml") }
    }

    @Test
    fun `returns empty list for feed with no items`() {
        val emptyFeed = """<?xml version="1.0"?>
            <rss version="2.0" xmlns:sparkle="http://www.andymatuschak.org/xml-namespaces/sparkle">
              <channel/>
            </rss>"""
        val items = withOs("windows") { parser.parse(emptyFeed) }
        assertTrue(items.isEmpty())
    }

    @Test
    fun `ignores items missing sparkle version element`() {
        val feed = """<?xml version="1.0"?>
            <rss version="2.0" xmlns:sparkle="http://www.andymatuschak.org/xml-namespaces/sparkle">
              <channel>
                <item>
                  <title>No version</title>
                  <enclosure url="https://example.com/x.exe" sparkle:os="windows" length="100" type="application/octet-stream"/>
                </item>
              </channel>
            </rss>"""
        val items = withOs("windows") { parser.parse(feed) }
        assertTrue(items.isEmpty())
    }

    // --- helpers ---

    private fun <T> withOs(os: String, block: () -> T): T {
        val original = System.getProperty("os.name")
        System.setProperty(
            "os.name",
            when (os) {
                "windows" -> "Windows 10"
                "macos" -> "Mac OS X"
                else -> "Linux"
            },
        )
        return try {
            block()
        } finally {
            System.setProperty("os.name", original)
        }
    }
}
